icicle-compiler/src/Icicle/Sea/FromAvalanche/Base.hs

```diff
@@ -35,25 +40,75 @@ import           Text.Printf (printf)
 
 ------------------------------------------------------------------------
 
-textOfName :: Pretty n => n -> Text
-textOfName = T.pack . show . seaOfName
-
-seaOfName :: Pretty n => n -> Doc
-seaOfName = string . fmap mangle . show . pretty
-  where
-    mangle '$' = '_'
-    mangle ' ' = '_'
-    mangle  c  =  c
-
-seaOfNameIx :: Pretty n => n -> Int -> Doc
-seaOfNameIx n ix = seaOfName (pretty n <> text "$ix$" <> int ix)
-
-seaOfAttributeDesc :: Attribute -> Doc
-seaOfAttributeDesc (Attribute xs)
-  | T.null xs = string ""
-  | otherwise = pretty (T.filter isLegal xs)
- where
-  isLegal c = isLower c || isUpper c || c == ' ' || c == '_'
+-- | A legal C identifier.
+newtype SeaName = SeaName {
+    getSeaName :: Text
+ } deriving (Eq, Ord, Show)
+
+-- | A value that will need be quoted in C.
+newtype SeaString = SeaString {
+    getSeaString :: Text
+ } deriving (Eq, Ord, Show)
+
+takeSeaName :: SeaName -> Text
+takeSeaName = getSeaName
+
+takeSeaString :: SeaString -> Text
+takeSeaString = getSeaString
+
+seaNameValidHead :: Char -> Bool
+seaNameValidHead c =
+  (c >= 'a' && c <= 'z') ||
+  (c >= 'A' && c <= 'Z')
+{-# INLINE seaNameValidHead #-}
+
+seaNameValidTail :: Char -> Bool
+seaNameValidTail c =
+  seaNameValidHead c ||
+  (c >= '0' && c <= '9') ||
+  (c == '_')
+{-# INLINE seaNameValidTail #-}
+
+asSeaName :: Text -> Maybe SeaName
+asSeaName t =
+  case Text.unpack t of
+    x:xs | seaNameValidHead x && all seaNameValidTail xs ->
+      Just (SeaName t)
+    _ ->
+      Nothing
+
+mangleToSeaName :: Pretty n => n -> SeaName
+mangleToSeaName (show . pretty -> n) =
+  case n of
+    [] ->
+      -- Technically not right as "" is not a legal C identifier.
+      -- This is just how we chose to treat "".
+      SeaName ""
+    x:_  ->
+      let
+        n'
+          | isAlpha x = n
+          | otherwise = "x_" <> n
+        mangle c
+          | c >= '0' && c <= '9'
+          = [c]
+          | c >= 'a' && c <= 'z'
+          = [c]
+          | c >= 'A' && c <= 'Z'
+          = [c]
+          | c == '_'
+          = "__"
+          | otherwise
+          = "u_" <> showHex (ord c) ""
```

Comment:
do we need '_' suffix as well, to signal the end of hex, in case next char is digit/hex?